src/client/mod.rs

```diff
@@ -0,0 +1,198 @@
+use crate::transport::Transport;
+use crate::Result;
+
+#[cfg(target_arch = "wasm32")]
+use crate::transport::WasmClient;
+
+#[cfg(not(target_arch = "wasm32"))]
+use crate::transport::HyperClient;
+
+mod builder;
+pub use builder::ClientBuilder;
+
+#[allow(dead_code)]
+pub const TYPESENSE_API_KEY_HEADER_NAME: &str = "X-TYPESENSE-API-KEY";
+
+/// Root client for top level APIs
+pub struct Client<'a, T> {
+    transport: Transport<T>,
+    host: &'a str,
+    api_key: &'a str,
+}
+
+impl<'a, T> Client<'a, T> {
+    /// Gets the transport of the client
+    pub fn transport(&self) -> &Transport<T> {
+        &self.transport
+    }
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+#[allow(dead_code)]
+impl<'a, C> Client<'a, HyperClient<C>>
+where
+    C: hyper::client::connect::Connect + Clone + Send + Sync + 'static,
+{
+    pub(crate) async fn send(
+        &self,
+        method: http::Method,
+        path: &str,
+        mut headers: http::HeaderMap,
+        body: Vec<u8>,
+    ) -> Result<http::Response<hyper::Body>> {
+        let uri = format!("{}{}", self.host, path);
+        headers.insert(TYPESENSE_API_KEY_HEADER_NAME, self.api_key.parse().unwrap());
+        self.transport.send(method, &uri, headers, body).await
+    }
+
+    pub(crate) async fn get(
+        &self,
+        path: &str,
+        headers: http::HeaderMap,
+    ) -> Result<http::Response<hyper::Body>> {
+        self.send(http::Method::GET, path, headers, vec![]).await
+    }
+
+    pub(crate) async fn post(
+        &self,
+        path: &str,
+        headers: http::HeaderMap,
+        body: Vec<u8>,
+    ) -> Result<http::Response<hyper::Body>> {
+        self.send(http::Method::POST, path, headers, body).await
+    }
+}
+
+#[cfg(target_arch = "wasm32")]
+#[allow(dead_code)]
+impl<'a> Client<'a, WasmClient> {
+    pub(crate) async fn send(
+        &self,
+        method: http::Method,
+        path: &str,
+        mut headers: http::HeaderMap,
+        body: Vec<u8>,
+    ) -> Result<http::Response<Vec<u8>>> {
+        let uri = format!("{}{}", self.host, path);
+        headers.insert(TYPESENSE_API_KEY_HEADER_NAME, self.api_key.parse().unwrap());
+        self.transport.send(method, &uri, headers, body).await
+    }
+
+    pub(crate) async fn get(
+        &self,
+        path: &str,
+        headers: http::HeaderMap,
+    ) -> Result<http::Response<Vec<u8>>> {
+        self.send(http::Method::GET, path, headers, vec![]).await
+    }
+
+    pub(crate) async fn post(
+        &self,
+        path: &str,
+        headers: http::HeaderMap,
+        body: Vec<u8>,
+    ) -> Result<http::Response<Vec<u8>>> {
+        self.send(http::Method::POST, path, headers, body).await
+    }
+}
```

Comment:
Not a top priority now, but I thought of changing the HttpLowLevel trait to something like this:

---

```rust
trait HttpLowLevel<M = http::Method, H = http::HeaderMap> {
    // ...
}
```

---

This might allow us to use generics here, and reduce the implementation for different types.

Reply:
Yes, I also noticed that in our implementation the only difference is in the response type. We coud also remove from HttpLowLevel the method, headermap and body type and assume that we are always using the http::* structs.

Reply:
Update: changed  HttpLowLevel to HttpLowLevel<M = http::Method, H = http::HeaderMap, B = Vec<u8>> {